### PR TITLE
Manual section on prompt buffer.

### PR DESCRIPTION
**Pull request by aadcg**
A follow up of the manual section drafted in #2655. 

# Description

Work in progress.  I'll let you know when to take a look @aartaka.

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.

- [ ] I have pulled from master before submitting this PR
- [ ] There are no merge conflicts.
- [ ] I've added the new dependencies as:
  - [ ] ASDF dependencies,
  - [ ] Git submodules,
    ```sh
	cd /path/to/nyxt/checkout
    git submodule add https://gitlab.common-lisp.net/nyxt/py-configparser _build/py-configparser
    ```
  - [ ] and Guix dependencies.
- [ ] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [ ] I have performed a self-review of my own code.
- [ ] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [ ] Documentation:
  - [ ] All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - [ ] I have updated the existing documentation to match my changes.
  - [ ] I have commented my code in hard-to-understand areas.
  - [ ] I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
  - [ ] I have added a `migration.lisp` entry for all compatibility-breaking changes.
  - [ ] (If this changes something about the features showcased on Nyxt website) I have these changes described in the new/existing article at Nyxt website or will notify one of maintainters to do so.
- [ ] Compilation and tests:
  - [ ] My changes generate no new warnings.
  - [ ] I have added tests that prove my fix is effective or that my feature works.  (If possible.)
  - [ ] New and existing unit tests pass locally with my changes.

### CODE REVIEW

**Comment by aadcg** [comment]
I am now backporting [an article](https://nyxt.atlas.engineer/article/prompt-buffer-customization.org) written by @Ambrevar which serves as the base to a section called "Programming Interface".

**Comment by aadcg** [comment]
Here is the first draft of the prompt buffer manual section (still in `.org` mode so that editing is easier, it will be converted to our format at the last minute).

At this point we should validate the following:

- [ ] The logical ordering of the sections;
- [ ] The amount of information - what is missing or extra?; 
- [ ] The clarity of each section

Broadly there are two distinct sections in the document. In the first part, the target is the user. In the second, (introduced by "Programming Interface") the target is the developer.

Ideally everyone should read it. CC @aartaka @Ambrevar @jmercouris.

**Comment by Ambrevar** [comment]
To clarify, do you want a review now or when you will be done with the draft?

**Comment by aadcg** [comment]
@Ambrevar  now, the draft is ready to review.

File: source/manual-prompt-buffer.org [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+next-suggestion
+previous-suggestion
+first-suggestion
+last-suggestion
+next-page
+previous-page
+next-source
+previous-source
+
+Oftentimes, you will interact with a collection of suggestions, which we call
```
Comment (aartaka) [inline source/manual-prompt-buffer.org:33]: Maybe mention why we use marks—to act on more than one suggestion at a time?

File: source/manual-prompt-buffer.org [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+top.
+
+The mechanism is flexible enough to guess what you mean.  For instance, if you
+type "buffer make" it is able to understand that you're looking for the command
+make-buffer.
+
+Notice that the contents of the input bar is matched against all suggestion
+attributes shown.
+
+The syntax of regular expressions is not supported.
```
Comment (aartaka) [inline source/manual-prompt-buffer.org:115]: "... but configurable"?

File: source/manual-prompt-buffer.org [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+** Programming Interface
+# Loosely based on:
+# https://nyxt.atlas.engineer/article/prompt-buffer-customization.org
+The system nyxt/prompter (containing the :prompter package) defines the
+primitives that make prompt buffers possible in Nyxt.
+
+Within the :nyxt package, a set of primitives is defined on top of :prompter to
+suit the browser's needs.  Additionally, the package :nyxt/prompt-buffer-mode
+defines the relevant mode.
```
Comment (aartaka) [inline source/manual-prompt-buffer.org:125]: What does mode do? It's worth mentioning that it hosts necessary commands and keybindings.

File: source/manual-prompt-buffer.org [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+Much of the prompt buffer's flexibility owes to the options provided when
+defining sources.  Each source can be tailored to specific cases, since besides
+setting the suggestions accordingly, other parameters such as actions or filters
+are available.
+
+A source is an object of class prompter:source, or inheriting from it.  Let's
+define an example-source:
+
+#+begin_src lisp
+(defun set-suggestions (n)
```
Comment (aartaka) [inline source/manual-prompt-buffer.org:159]: This additional function may be quite confusing. I'd rather use something simpler (inline constructor?), at least for the first appearance of sources.

File: source/manual-prompt-buffer.org [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+You can further study it by looking at prompter:source.
+
+With the class defined above, the prompt buffer can be invoked as follows:
+
+#+begin_src lisp
+(prompt :prompt "Prompter"
+        :sources (make-instance 'example-source))
+#+end_src
+
+Sources can be inspected with the describe-class command.  Particularly
```
Comment (aartaka) [inline source/manual-prompt-buffer.org:186]: Dunno whether this `describe-class` reference belong here. But the source listing, augmented with `:nxref` links is a worthy thing.

File: source/manual-prompt-buffer.org [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+*** Actions
+# Follow outcome of https://github.com/atlas-engineer/nyxt/issues/2794
+Actions are set via the following slots:
+
+- =prompter:actions-on-return=
+- =prompter:actions-on-current-suggestion=
+- =prompter:actions-on-marks=
+
+An action is either a symbol or a command, that may be locally-defined by either
+=lambda-mapped-command= or =lambda-command=.
```
Comment (aartaka) [inline source/manual-prompt-buffer.org:199]: `lambda-unmapped-command` is useful too.

File: source/manual-prompt-buffer.org [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+#+end_src
+
+*** Attributes
+Suggestion values are arbitrary objects: strings, numbers, URLs, structures or
+class instances.  Particularly for the last two, it's convenient to display some
+of its slots.
+
+The =prompter:object-attributes= method serves precisely that purpose.  It
+enables displaying something more meaningful than the standard printed
+representation of class instance =#<CLASS-NAME {ID}>=.
```
Comment (aartaka) [inline source/manual-prompt-buffer.org:228]: Note that this representation is not really standardized beyond the sharp-less-than syntax. SBCL, CCL, ECL, and other implementations have it slightly different.

SBCL:
```
#<FOO {100833DF63}>
```

CCL:
```
#<FOO #x302000F91BBD>
```

ECL:
```
#<a COMMON-LISP-USER::FOO 0x7fbe8c6e62c0>
```

File: source/manual-prompt-buffer.org [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+
+This per-suggestion stepping allows the source view to update its display at
+regular intervals until all suggestions are filtered.
+
+This filter function is particularly useful to score suggestions by relevance.
+
+Finally, the =prompter:filter-postprocessor= is run of the result of the filter,
+this time again over the entire list at once.
+
+Once again, the source view is only updated when the postprocessor is done with
```
Comment (aartaka) [inline source/manual-prompt-buffer.org:264]: Mention that postprocessor is asynchronous, in contrast to preprocessor?

File: source/manual-prompt-buffer.org [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+best.
+
+*** Suggestion objects
+Internally, when the prompter filters and sorts suggestions, it wraps the list
+of initial values (arbitrary objects) into =suggestion= objects.
+
+The class is useful to store information through the processing pipeline
+described in the previous section.  For instance, the filter can calculate and
+store the score of each suggestion, in the =score= slot, which in turn is
+readily available to the =filter-postprocessor= for sorting.
```
Comment (aartaka) [inline source/manual-prompt-buffer.org:283]: Mention the `prompter:suggestion-maker` here and add an example code to redefine it?

**Review by aartaka — APPROVED** [review]

File: source/manual-prompt-buffer.org [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+Generally speaking, the filtering mechanism is set to fuzzy matching which
+orders suggestions in such a way that the most relevant ones will appear at the
+top.
+
+The mechanism is flexible enough to guess what you mean.  For instance, if you
+type "buffer make" it is able to understand that you're looking for the command
+make-buffer.
+
+Notice that the contents of the input bar is matched against all suggestion
+attributes shown.
```
Comment (Ambrevar) [inline source/manual-prompt-buffer.org:113]: I don't understand this sentence.

File: source/manual-prompt-buffer.org [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+
+The syntax of regular expressions is not supported.
+
+** Programming Interface
+# Loosely based on:
+# https://nyxt.atlas.engineer/article/prompt-buffer-customization.org
+The system nyxt/prompter (containing the :prompter package) defines the
+primitives that make prompt buffers possible in Nyxt.
+
+Within the :nyxt package, a set of primitives is defined on top of :prompter to
```
Comment (Ambrevar) [inline source/manual-prompt-buffer.org:123]: What is `:prompter`?

File: source/manual-prompt-buffer.org [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+  (:export-class-name-p t)
+  (:metaclass user-class))
+#+end_src
+
+Slots =prompter:name= and =prompter:constructor= are required.
+
+# Link to slots.
+# (:nxref :slot 'prompter:constructor :class-name 'prompter:source)
+
+You can further study it by looking at prompter:source.
```
Comment (Ambrevar) [inline source/manual-prompt-buffer.org:177]: Inspect prompter:source for more details.

File: source/manual-prompt-buffer.org [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+
+With the class defined above, the prompt buffer can be invoked as follows:
+
+#+begin_src lisp
+(prompt :prompt "Prompter"
+        :sources (make-instance 'example-source))
```
Comment (Ambrevar) [inline source/manual-prompt-buffer.org:183]: You can show the full form with `:sources (list (make-instance ...))` and then show the shortcut with neither `list` nor `make-instance`.

File: source/manual-prompt-buffer.org [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+Actions are set via the following slots:
+
+- =prompter:actions-on-return=
+- =prompter:actions-on-current-suggestion=
+- =prompter:actions-on-marks=
+
+An action is either a symbol or a command, that may be locally-defined by either
+=lambda-mapped-command= or =lambda-command=.
+
+Regardless, actions are functions that takes a list as a sole argument - either
```
Comment (Ambrevar) [inline source/manual-prompt-buffer.org:201]: typo: "that take"

File: source/manual-prompt-buffer.org [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+
+Then, the =prompter:filter= is run over the suggestions returned by the
+preprocessor, one after the other.
+
+This per-suggestion stepping allows the source view to update its display at
+regular intervals until all suggestions are filtered.
+
+This filter function is particularly useful to score suggestions by relevance.
+
+Finally, the =prompter:filter-postprocessor= is run of the result of the filter,
```
Comment (Ambrevar) [inline source/manual-prompt-buffer.org:261]: "on the result"?

**Comment by jmercouris** [comment]
@aadcg can I review and merge this?

**Comment by jmercouris** [comment]
If you would like to use org-mode, we would lose /a lot/ of the abilities we have right now. I think our documentation has become a lot more rich because of embedding it directly into the source code. The only thing I might suggest, that *might* make writing documentation easier would be an embeddable DSL. That is however a large project, and a discussion for a different time.

**Comment by jmercouris** [comment]
I'm sorry, it was again an accident, the stupid page froze on my machine and I don't even know what it clicked or didn't. I meant to only close the PR, not delete the branch! I just wanted to remove clutter and keep only the fresh active things in view.

**Pull request closed without merging**